File: data_handler.py
import connection
from psycopg2 import sql
import bcrypt


@connection.connection_handler
def get_boards(cursor):
    cursor.execute("""
                    SELECT * FROM boards
    """)
    boards = cursor.fetchall()
    return boards


@connection.connection_handler
def get_cards(cursor, board_id, status_id):
    cursor.execute("""
                    SELECT * FROM cards
                    WHERE board_id=%(board_id)s AND status_id=%(status_id)s;
    """,
                   {"board_id": board_id, "status_id": status_id})
    cards = cursor.fetchall()
    return cards


@connection.connection_handler
def get_one_card(cursor, card_id):
    cursor.execute("""
                SELECT * FROM cards
                WHERE id=%(card_id)s;
    """,
                   {"card_id": card_id})
    card = cursor.fetchone()
    return card


@connection.connection_handler
def change_status(cursor, card_id, status_id):
    cursor.execute("""
                UPDATE cards
                SET status_id=%(status_id)s
                WHERE id=%(card_id)s;
    """,
                   {"card_id": card_id, "status_id": status_id})


@connection.connection_handler
def create_new_board(cursor):
    cursor.execute('''
                        SELECT MAX(id) from boards
                        ''')
    seq = cursor.fetchone()
    seq = seq['max'] + 1

    cursor.execute('''
                        INSERT INTO boards (title) VALUES (%(seq)s)
                        ''',
                   {'seq': seq})


@connection.connection_handler
def create_new_card(cursor, board_id, status_id):
    cursor.execute("""
                        INSERT INTO cards (board_id, title, status_id, orderd)
                        VALUES (%(board_id)s, 'New Card', %(status_id)s, 0)
    """,
                   {'board_id': board_id, 'status_id': status_id})


@connection.connection_handler
def get_statuses(cursor):
    cursor.execute("""
                        SELECT * from statuses 
        """)
    statuses = cursor.fetchall()
    return statuses


@connection.connection_handler
def delete_board(cursor, board_id):
    cursor.execute('''
                        DELETE FROM  boards
                        WHERE id = %(board_id)s;
                        ''',
                   {'board_id': board_id})

@connection.connection_handler
def delete_card(cursor, card_id):
    cursor.execute('''
                        DELETE FROM cards
                        WHERE id = %(card_id)s;
                        ''',
                   {'card_id': card_id})

File: static/js/data_handler.js
// this object contains the functions which handle the data and its reading/writing
// feel free to extend and change to fit your needs

// (watch out: when you would like to use a property/function of an object from the
// object itself then you must use the 'this' keyword before. For example: 'this._data' below)
export let dataHandler = {
    _data: {}, // it contains the boards and their cards and statuses. It is not called from outside.
    _api_get: function (url, callback) {
        // it is not called from outside
        // loads data from API, parses it and calls the callback with it

        fetch(url, {
            method: 'GET',
            credentials: 'same-origin'
        })
            .then(response => response.json())  // parse the response as JSON
            .then(json_response => callback(json_response));  // Call the `callback` with the returned object
    },
    _api_post: function (url, data, callback) {
        // it is not called from outside
        // sends the data to the API, and calls callback function
        fetch(url, {
            method: 'POST',
            credentials: 'same-origin',
            body: JSON.stringify(data),
        })
            .then(response => response.json())  // parse the response as JSON
            .then(json_response => callback(json_response));  // Call the `callback` with the returned object
    },
    init: function () {
    },
    getBoards: function (callback) {
        // the boards are retrieved and then the callback function is called with the boards

        // Here we use an arrow function to keep the value of 'this' on dataHandler.
        //    if we would use function(){...} here, the value of 'this' would change.
        this._api_get('/get-boards', (response) => {
            this._data = response;
            callback(response);
        });
    },
    getBoard: function (boardId, callback) {
        // the board is retrieved and then the callback function is called with the board
        this._api_get(`/get-boards/${boardId}`, (response) => {
            this._data = response;
            callback(response);
        });
    },
    getStatuses: function (callback) {
        // the statuses are retrieved and then the callback function is called with the statuses
        this._api_get('/get-statuses', (response) => {
            this._data = response;
            callback(response);
        })
    },
    getStatus: function (statusId, callback) {
        // the status is retrieved and then the callback function is called with the status
    },
    getCardsByBoardId: function (boardId, statusId, callback) {
        // the cards are retrieved and then the callback function is called with the cards
        this._api_get(`/get-cards/${boardId}/${statusId}`, (response) => {
            this._data = response;
            callback(response);
        });
    },
    getCard: function (cardId, callback) {
        // the card is retrieved and then the callback function is called with the card
    },
    createNewBoard: function (callback) {
        // creates new board, saves it and calls the callback function with its data
        this._api_get('/create-new-board', (response) => {
            this._data = response;
            callback(response);
        });
    },
    createNewCard: function (boardId, statusId, callback) {
        // creates new card, saves it and calls the callback function with its data
        this._api_get(`/create-new-card/${boardId}/${statusId}`, (response) => {
            this._data = response;
            callback(response)
        })
    },
    deleteBoard: function (boardId, callback) {
        // deletes board, saves it and calls the callback function with its data
        this._api_get(`/delete-board/${boardId}`, (response) => {
            this._data = response;
            callback(response);
        });
    },
    deleteCard: function (cardId, callback) {
        // deletes board, saves it and calls the callback function with its data
        this._api_get(`/delete-card/${cardId}`, (response) => {
            this._data = response;
            callback(response);
        });
        // here comes more features
    },

    renameBoard: function (boardId, callback) {
        this._api_get(`/rename-board/${boardId}`, (response) => {
            this._data = response;
            callback(response);
        })
    }
};
